ouroboros-consensus/docs/AddingAnEra.md

```diff
@@ -0,0 +1,126 @@
+  until the required instances have been added. This means that we can't extend
+  the library with the new era without extending the test suite too. In the
+  past, this has been the main blocker: the library was ready, but the required
+  instances in the ledger were missing for our test suite(s).
+
+* Add the required dependencies on the ledger package (and its test package) to
+  the `cabal.project` file. You will have to add it to some other cabal files
+  too, let the compiler tell you which ones.
+
+## `ouroboros-consensus-shelley`
+
+* Define `StandardAlonzo` in `Ouroboros.Consensus.Shelley.Eras` and add any
+  missing instances to that module, update the export list appropriately.
+
+* In `Ouroboros.Consensus.Shelley.Node`, define `ProtocolParamsAlonzo` just like
+  the existing ones. In case the era in question needs extra parameters (e.g., a
+  new genesis config?), they can be added here.
+
+## `ouroboros-consensus-shelley-test`
+
+* In `Test.Consensus.Shelley.Examples`, define an `examplesAlonzo` similar to
+  how `examplesMary` is defined. If necessary, i.e., when new type families have
+  been created, the `examples` function might have to take more arguments. This
+  is where the golden test examples are defined, but only the Shelley ones are
+  tested as part of this test suite. The others are only tested as part of the
+  `ouroboros-consensus-cardano-test` test suite.
+
+## `ouroboros-consensus-cardano`
+
+* In `Ouroboros.Consensus.Cardano.Block`, include `AlonzoEra` in `CardanoEras`
+  and `ShelleyBasedEras`. Update all the pattern synonyms in the module with the
+  new era. Don't forget to update the comments, the `COMPLETE` pragmas, and the
+  export lists. It's easy to forget a case and the compiler will likely not warn
+  you, you'll notice it when trying to use the pattern synonyms.
+
+* In `Ouroboros.Consensus.Cardano.CanHardFork`, update
+  `CardanoHardForkConstraints`, add additional translations to the `CanHardFork`
+  instance.
+
+* In `Ouroboros.Consensus.Cardano.Node`, update the `SerialiseHFC` instance by
+  following the existing patterns. Add a new `CardanoNodeToNodeVersion` and
+  `CardanoNodeToClientVersion` that enable the `AlonzoEra`, update the existing
+  ones so that they disable the new era. Be sure to include the new versions in
+  the two methods of the `SupportedNetworkProtocolVersion` instance. Extend
+  `protocolInfoCardano` with the new era by following the type errors and adding
+  the missing parameters (including `ProtocolParamsTransition`). Don't forget to
+  derive `maxMajorProtVer` from the new final era. Update
+  `protocolClientInfoCardano` too.
+
+* In `Ouroboros.Consensus.Cardano`, update the `ProtocolCardano` type synonym,
+  add the extra arguments needed for `protocolInfoCardano` to the
+  `ProtocolCardano` constructor. Update `protocolInfo` accordingly.
+
+## `ouroboros-consensus-cardano-test`
+
+* In `Test.Consensus.Cardano.Generators`, update `arbitraryNodeToNode`,
```

Comment:
This was the first module where everything felt unfamiliar and it wasn't "obvious" to me how to follow your instructions. It's also where I ran out of time for today. I'll pick up here next time.

Reply:
Hmm, I could do some more hand holding here, but it's mostly a matter of following the existing patterns. After you have picked up from here, we can reconsider.

Reply:
OK. I understand the pattern now. Especially if I just search for the name of a previous era, it seems easy enough to handle this module.

I think my hesitation is do mostly to feeling lost in this sea of instances in this module. I don't think there's much to be done there in advance of me spending to the time to understand them all.

Unless you can think of a pithy Haddock comment for the module header? Something about serialization tests, IIUC.

Reply:
I added a Haddock comment for the module header.